### PR TITLE
feat: support the processing of GetContinuousQueryLease and assignment of CQ tasks

### LINKED ISSUE #308: support task assignment and status report of cq
### Is your feature request related to a problem? Please describe.（如果该特性或者功能和已知某个问题相关，请描述问题是什么）

1. ts-sql needs to periodically fetch cqs from ts-meta to execute or stop cq.
2. ts-sql needs to periodically send heartbeats to ts-meta to prove its survival.
3. The last run time of cq needs to be sent to ts-meta for persistence.

### Describe the solution you'd like（问题的解决办法是什么）

The existing designs are as follows:
1. ts-sql periodically sends heartbeats to ts-meta, reports its own survival.
2. ts-meta records the relevant information of ts-sql according to the heartbeat of ts-sql, but does not need to be persisted.
3. In the meta data of CQ, mark the ts-sql of the CQ that should be executed, so that ts-sql knows which CQs need to be executed and which ones need to be stopped after getting the meta data.
4. When ts-sql receives the task of executing cq, it starts a goroutine to execute cq, and after completing a `select into` operation, reports the last run time to ts-meta for persistence.
5. When ts-sql receives the task of stopping cq, it closes the corresponding goroutine.
6. Store all cqs running on each ts-sql in ts-meta for scheduling and load balancing.

### Describe alternatives you've considered（描述您考虑过的替代方案）

No alternatives.

### Additional context（其他补充内容）

The scheduling and load balancing of cq can be further improved on this basis.

### PR DESCRIPTION
**Pull request by Benevor**
### What problem does this PR solve?

Issue Number: close #308

### What is changed and how it works?
Implement `GetContinuousQueryLease` RPC message. In CQ service, every time the `handle()` is called, the lease needs to be updated first. When ts-meta processes the first getlease request of ts-sql, all cq tasks should be treated as responses, rather than just responding to changing tasks.
In this scheme, only when creating a cq, deleting a cq, adding a new ts-sql, and ts-sql fails to heartbeat for a long time, the cq scheduling is triggered.

### How Has This Been Tested?

I have added unit tests to the corresponding files to verify the correctness of the `GetContinuousQueryLeaseCommand ` processing results. And I verified the correctness of `GetContinuousQueryLeaseCommand ` sending and processing by manually debugging the program.
- [x] full build of project.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/ts-meta/meta/sql_node_info.go [hunk 1]
```diff
@@ -21,10 +21,27 @@ import (
...
 }
 
 type HeartbeatInfo struct {
 	Host              string
 	LastHeartbeatTime time.Time
 }
+
+type CqInfo struct {
+	// all cqs that running on this ts-sql
+	Cqs map[string]string
```
Comment (shilinlee) [inline app/ts-meta/meta/sql_node_info.go:34]:
```suggestion
	RunningCqs map[string]string
```

File: app/ts-meta/meta/store.go [hunk 2]
```diff
@@ -1391,18 +1393,16 @@ func (s *Store) applySql2MetaHeartbeat(host string) error {
 	if !s.IsLeader() {
 		return raft.ErrNotLeader
 	}
-	s.sqlMu.RLock()
-	defer s.sqlMu.RUnlock()
+	s.sqlMu.Lock()
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:1396]: The write lock scope is too broad.

File: app/ts-meta/meta/store.go [hunk 3]
```diff
@@ -400,6 +400,7 @@ type Store struct {
 	sqlMu             sync.RWMutex            // concurrency for heartbeat processing and crash detection
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:400]: Please check the granularity of the lock.

File: app/ts-meta/meta/store.go [hunk 4]
```diff
@@ -1434,38 +1438,161 @@ func (s *Store) detectSqlNodeCrash() {
...
+			continue
+		}
+		if sni.CqInfo.GetLoad() < num {
+			sql = s
+		}
+	}
+	s.SqlNodes[sql].CqInfo.AssignCqs[cq] = cq
+	if s.SqlNodes[sql].CqInfo.IsNew {
+		s.SqlNodes[sql].CqInfo.IsNew = false
+	}
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:1516]:
```suggestion
        s.SqlNodes[sql].CqInfo.IsNew = false
```

File: app/ts-meta/meta/store.go [hunk 5]
```diff
@@ -1434,38 +1438,161 @@ func (s *Store) detectSqlNodeCrash() {
...
+	delete(s.SqlNodes, sqlName)
+	// delete HeartbeatInfo.
+	s.HeartbeatInfoList.Remove(hbi)
+
+	for _, cq := range cqs {
+		s.findSqlByCq(cq)
+	}
+}
+
+func (s *Store) findSqlByCq(cq string) {
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:1500]:
```suggestion
func (s *Store) assignCQ2SQL(cq string) {
```

File: app/ts-meta/meta/store.go [hunk 6]
```diff
@@ -400,6 +400,7 @@ type Store struct {
 	sqlMu             sync.RWMutex            // concurrency for heartbeat processing and crash detection
 	SqlNodes          map[string]*SqlNodeInfo // all alive ts-sql host; no need for persistence
 	HeartbeatInfoList *list.List              // the latest heartbeat information for each ts-sql
+	Cq2Sql            map[string]string
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:403]:
```suggestion
	Cq2Sql            map[string]string // {"select * into ...": "127.0.0.1:8086"}
```

File: app/ts-meta/meta/store.go [hunk 7]
```diff
@@ -1434,38 +1438,161 @@ func (s *Store) detectSqlNodeCrash() {
...
+	s.HeartbeatInfoList.Remove(hbi)
+
+	for _, cq := range cqs {
+		s.findSqlByCq(cq)
+	}
+}
+
+func (s *Store) findSqlByCq(cq string) {
+	sql := ""
+	num := 0
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:1502]:
```suggestion
	var sql string
	var minLoad = math.MaxInt
```

File: lib/metaclient/meta_client.go [hunk 8]
```diff
@@ -685,6 +686,65 @@ func (c *Client) GetMeasurementInfoStore(dbName string, rpName string, mstName s
...
+func (c *Client) GetCqLease(firstTime bool, host string, cqs []string) ([]string, []string, error) {
+	val := &proto2.GetContinuousQueryLeaseCommand{
+		Host:      &host,
+		Cqs:       cqs,
+		FirstTime: &firstTime,
+	}
+	t := proto2.Command_GetContinuousQueryLeaseCommand
+	cmd := &proto2.Command{Type: &t}
+	if err := proto.SetExtension(cmd, proto2.E_GetContinuousQueryLeaseCommand_Command, val); err != nil {
+		panic(err)
```
Comment (shilinlee) [inline lib/metaclient/meta_client.go:743]: What does this code snippet do?

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-meta/meta/store.go [hunk 9]
```diff
@@ -397,9 +397,11 @@ type Store struct {
 	cacheDataBytes   []byte
 	cacheDataChanged chan struct{}
 
-	sqlMu             sync.RWMutex            // concurrency for heartbeat processing and crash detection
+	sqlMu             sync.RWMutex            // lock for SqlNodes
 	SqlNodes          map[string]*SqlNodeInfo // all alive ts-sql host; no need for persistence
+	hbMu              sync.RWMutex            // lock for HeartbeatInfoList
 	HeartbeatInfoList *list.List              // the latest heartbeat information for each ts-sql
+	// Cq2Sql            map[string]string       // {"select * into ...": "127.0.0.1:8086"}
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:404]:
```suggestion
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: lib/metaclient/meta_client.go (hunk 8 above)
Comment (Benevor) [reply lib/metaclient/meta_client.go]: I have deleted this code.

File: app/ts-meta/meta/message/message.codec.gen.go [hunk 10]
```diff
@@ -924,3 +924,71 @@ func (o *Sql2MetaHeartbeatResponse) Size() int {
...
+	return &GetContinuousQueryLeaseRequest{}
+}
+
+func (o *GetContinuousQueryLeaseResponse) Marshal(buf []byte) ([]byte, error) {
+	var err error
+	buf = codec.AppendStringSlice(buf, o.AssignCqs)
+	buf = codec.AppendStringSlice(buf, o.RevokeCqs)
+	buf = codec.AppendString(buf, o.Err)
+
+	return buf, err
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/message/message.codec.gen.go]: I have fixed it.

File: app/ts-meta/meta/message/message.codec.gen.go [hunk 11]
```diff
@@ -924,3 +924,71 @@ func (o *Sql2MetaHeartbeatResponse) Size() int {
...
+	if len(buf) == 0 {
+		return nil
+	}
+	var err error
+	dec := codec.NewBinaryDecoder(buf)
+	o.AssignCqs = dec.StringSlice()
+	o.RevokeCqs = dec.StringSlice()
+	o.Err = dec.String()
+
+	return err
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/message/message.codec.gen.go]: I have fixed it.

File: app/ts-meta/meta/store.go (hunk 9 above)
Comment (Benevor) [reply app/ts-meta/meta/store.go]: I have fixed it.

**Pull request merged**